src/front/glsl/ast.rs

```diff
@@ -257,64 +314,209 @@ impl Context {
     pub fn remove_current_scope(&mut self) {
         self.scopes.pop();
     }
+
+    pub fn lower(
+        &mut self,
+        program: &mut Program,
+        expr: Expr,
+        lhs: bool,
+        body: &mut Block,
+    ) -> Result<Handle<Expression>, ErrorKind> {
+        Ok(match expr.kind {
+            ExprKind::Access { base, index } => {
+                let base = self.lower(program, *base, lhs, body)?;
+                let index = self.lower(program, *index, false, body)?;
+
+                self.expressions.append(Expression::Access { base, index })
+            }
+            ExprKind::Select { base, field } => {
+                let base = self.lower(program, *base, lhs, body)?;
+
+                program.field_selection(self, base, &field, expr.meta)?
+            }
+            ExprKind::Constant(constant) if !lhs => {
+                self.expressions.append(Expression::Constant(constant))
+            }
+            ExprKind::Binary { left, op, right } if !lhs => {
+                let left = self.lower(program, *left, false, body)?;
+                let right = self.lower(program, *right, false, body)?;
+
+                if let BinaryOperator::Equal | BinaryOperator::NotEqual = op {
+                    let equals = op == BinaryOperator::Equal;
+                    let left_is_vector = match *program.resolve_type(self, left)? {
+                        crate::TypeInner::Vector { .. } => true,
+                        _ => false,
+                    };
+
+                    let right_is_vector = match *program.resolve_type(self, right)? {
+                        crate::TypeInner::Vector { .. } => true,
+                        _ => false,
+                    };
+
+                    let (op, fun) = match equals {
+                        true => (BinaryOperator::Equal, RelationalFunction::All),
+                        false => (BinaryOperator::NotEqual, RelationalFunction::Any),
+                    };
+
+                    let expr = self
+                        .expressions
+                        .append(Expression::Binary { op, left, right });
+
+                    if left_is_vector && right_is_vector {
```

Comment:
If only one of them is a vector, is this valid in GLSL? If it's valid, I assume we'd need to `Splat` the scalar one first?